File: frontend/src/types/job.ts
export type JobStatus = "pending" | "running" | "completed" | "failed";

export type ModuleName =
  | "uniprot"
  | "string"
  | "gprofiler"
  | "hpa"
  | "signalp"
  | "pharos"
  | "sasp"
  | "comparison"
  | "therapeutic"
  | "receptor_ligand"
  | "safety"
  | "disease_context";

export interface ModuleProgress {
  status: JobStatus;
  percent: number;
  message: string;
}

export interface Job {
  id: string;
  status: JobStatus;
  job_type: "single" | "comparison";
  proteins: string[];
  modules: string[];
  progress: Record<string, ModuleProgress>;
  error_message: string | null;
  label: string | null;
  proteins_a: string[] | null;
  proteins_b: string[] | null;
  set_a_label: string | null;
  set_b_label: string | null;
  created_at: string;
  updated_at: string;
}

export interface JobSummary {
  id: string;
  status: JobStatus;
  job_type: "single" | "comparison";
  label: string | null;
  set_a_label: string | null;
  set_b_label: string | null;
  created_at: string;
  updated_at: string;
}

export interface JobCreate {
  job_type?: "single" | "comparison";
  // single mode
  proteins?: string[];
  modules?: ModuleName[];
  label?: string;
  // comparison mode
  set_a_proteins?: string[];
  set_a_label?: string;
  set_b_proteins?: string[];
  set_b_label?: string;
}

File: backend/app/schemas/job.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

ModuleName = Literal[
    "uniprot", "string", "gprofiler", "hpa", "signalp", "pharos", "sasp", "comparison",
    "therapeutic", "receptor_ligand", "safety", "disease_context",
]

ALL_MODULES: list[str] = [
    "uniprot", "string", "gprofiler", "hpa", "signalp", "sasp",
    "therapeutic", "receptor_ligand", "safety", "disease_context",
]


class JobCreate(BaseModel):
    proteins: list[str] = Field(
        ..., min_length=1, max_length=1000, description="UniProt accession IDs"
    )
    modules: list[str] = Field(
        default_factory=lambda: list(ALL_MODULES),
        description="Which analysis modules to run",
    )
    label: str | None = Field(None, max_length=255)


class ModuleProgress(BaseModel):
    status: Literal["pending", "running", "completed", "failed"] = "pending"
    percent: int = 0
    message: str = ""


class JobRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    status: str
    proteins: list[str]
    modules: list[str]
    progress: dict[str, ModuleProgress] = {}
    error_message: str | None = None
    label: str | None = None
    created_at: datetime
    updated_at: datetime


class JobSummary(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    status: str
    label: str | None
    created_at: datetime
    updated_at: datetime
